ghutil/ghutil.go

```diff
@@ -423,94 +433,188 @@ func processPullRequest(ghc *GitHubClient, ctx context.Context, orgName string,
+			removeLabel(LabelClaExternal)
+		} else {
+			logging.Infof("  PR doesn't have [%s] label, and shouldn't", LabelClaExternal)
+			// Nothing to do here.
+		}
+	}
+
+	if pullRequestStatus.Compliant {
+		logging.Info("  PR is CLA-compliant")
+	} else {
+		logging.Info("  PR is NOT CLA-compliant:", pullRequestStatus.NonComplianceReason)
+	}
+
+	// Add or remove [cla: yes] and [cla: no] labels, as appropriate.
+	if pullRequestStatus.Compliant {
+		// if PR has [cla: no] label, remove it.
+		if issueClaLabelStatus.HasNo {
+			removeLabel(LabelClaNo)
+		} else {
+			logging.Infof("  No action needed: [%s] label already missing", LabelClaNo)
+		}
+		// if PR doesn't have [cla: yes] label, add it.
+		if !issueClaLabelStatus.HasYes {
+			if repoClaLabelStatus.HasYes {
 				addLabel(LabelClaYes)
-			} else {
-				logging.Infof("  No action needed: [%s] label already added", LabelClaYes)
 			}
-		} else /* !pullRequestIsCompliant */ {
-			labelsUpdatedForNonCompliance := false
-			// if PR doesn't have [cla: no] label, add it.
-			if !hasLabelClaNo {
+		} else {
+			logging.Infof("  No action needed: [%s] label already added", LabelClaYes)
+		}
+	} else /* !pullRequestIsCompliant */ {
+		shouldAddComment := false
+		// if PR doesn't have [cla: no] label, add it.
+		if !issueClaLabelStatus.HasNo {
+			if repoClaLabelStatus.HasNo {
 				addLabel(LabelClaNo)
-				labelsUpdatedForNonCompliance = true
-			} else {
-				logging.Infof("  No action needed: [%s] label already added", LabelClaNo)
 			}
-			// if PR has [cla: yes] label, remove it.
-			if hasLabelClaYes {
-				removeLabel(LabelClaYes)
-				labelsUpdatedForNonCompliance = true
-			} else {
-				logging.Infof("  No action needed: [%s] label already missing", LabelClaYes)
+			shouldAddComment = true
+		} else {
+			logging.Infof("  No action needed: [%s] label already added", LabelClaNo)
+		}
+		// if PR has [cla: yes] label, remove it.
+		if issueClaLabelStatus.HasYes {
+			removeLabel(LabelClaYes)
+			shouldAddComment = true
+		} else {
+			logging.Infof("  No action needed: [%s] label already missing", LabelClaYes)
+		}
+
+		if shouldAddComment {
+			addComment(pullRequestStatus.NonComplianceReason)
+		}
+	}
+
+	return nil
+}
+
+// IsExternal computes whether the given commit should be processed by this
+// tool, or if it should be covered by an external CLA management tool.
+func IsExternal(commit *github.RepositoryCommit, claSigners config.ClaSigners) bool {
+	if claSigners.External == nil {
+		return false
+	}
+
+	var usernames []string
+	if commit.Author != nil {
+		if commit.Author.Login != nil {
+			usernames = append(usernames, *commit.Author.Login)
+		}
+	}
+	if commit.Committer != nil {
+		if commit.Committer.Login != nil {
+			usernames = append(usernames, *commit.Committer.Login)
+		}
+	}
+
+	matchUsernames := func(usernames []string, accounts []config.Account) bool {
+		for _, account := range accounts {
+			for _, username := range usernames {
+				if username == account.Login {
```

Comment:
I was considering making a separate, limited container with just `Login` for external accounts, e.g.,

```go
type ExternalAccount struct {
	Login string
}
```

for when we want to not to manage the `Name`/`Email` (which we might not have) for externally-managed CLAs, but then code such as `matchUsernames()` which is designed to work for either type of an account doesn't generalize (I need a concrete type that's either `Account` or `ExternalAccount`), so I think it's just better to keep it as-is, unless you have a suggestion for how to enable code reuse while still removing the unused fields.

Thoughts?

Reply:
I thought this might be possible to do using an `interface` but interfaces can only have methods, not fields. There's a proposal to add fields to interfaces in Go v2: https://github.com/golang/go/issues/23796 but clearly, they're not available now, and I don't see how to simplify this code without it.